### PR TITLE
managed_addons fix

### PR DESCRIPTION
**Pull request by aayushsss1**
1. Diff Suppressed managed_addons parameter to prevent config change during plan and apply
2. Users can now set manage_all_addons to false after initial creation.

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

**Comment by hasan4791** [comment]
@hkantare PTAL. We've suppressed diff for the "managed_addons" parameter as it is a computed field and not set by users.

File: ibm/service/kubernetes/resource_ibm_container_addons.go [hunk 1]
```diff
@@ -528,12 +530,27 @@ func flattenAddOns(result []v1.AddOn) (resp *schema.Set, err error) {
...
+				var definedAddons []string
+				addOnSet := d.Get("addons").(*schema.Set).List()
+				for _, aoSet := range addOnSet {
+					ao, _ := aoSet.(map[string]interface{})
+					definedAddons = append(definedAddons, ao["name"].(string))
+				}
+				d.Set("managed_addons", flex.FlattenStringList(definedAddons))
+			}
+		}
 		managed_addons := d.Get("managed_addons").([]interface{})
```
Comment (hasan4791) [inline ibm/service/kubernetes/resource_ibm_container_addons.go:549]: This and line no:533 seems to be duplicate

File: ibm/service/kubernetes/resource_ibm_container_addons.go [hunk 2]
```diff
@@ -557,6 +574,20 @@ func resourceIBMContainerAddOnsUpdate(d *schema.ResourceData, meta interface{})
 			newList = new(schema.Set)
 		}
 		os := oldList.(*schema.Set)
+		// manage_all_addons was not previously set to false and the user has added additional or upgraded the existing addons, we set managed_addons to hold the old values
+		if !manageAllAddons && len(managed_addons) == 0 {
```
Comment (hasan4791) [inline ibm/service/kubernetes/resource_ibm_container_addons.go:578]: This condition would never hit considering the same condition at line no:536 at the start of this function!!

File: ibm/service/kubernetes/resource_ibm_container_addons.go [hunk 3]
```diff
@@ -58,10 +58,12 @@ func ResourceIBMContainerAddOns() *schema.Resource {
 				Description: "To manage all add-ons installed in the cluster using terraform by importing it into the state file",
 			},
 			"managed_addons": {
-				Type:        schema.TypeList,
-				Computed:    true,
-				Elem:        &schema.Schema{Type: schema.TypeString},
-				Description: "Used to keep track of the add-on names",
+				Type:             schema.TypeList,
```
Comment (hkantare) [inline ibm/service/kubernetes/resource_ibm_container_addons.go:61]: Why is computed variable change to optional?

File: ibm/service/kubernetes/resource_ibm_container_addons.go [hunk 4]
```diff
@@ -528,12 +530,27 @@ func flattenAddOns(result []v1.AddOn) (resp *schema.Set, err error) {
...
 func resourceIBMContainerAddOnsUpdate(d *schema.ResourceData, meta interface{}) error {
 	manageAllAddons := d.Get("manage_all_addons").(bool)
+	managed_addons := d.Get("managed_addons").([]interface{})
 	var addOnName []string
+	// If the user wants terraform to manage the defined addons
 	if !manageAllAddons {
+		// Case 1 - User has requested to manage addons after terraform creation and initialisation, but there is no change in the addon configuration
+		if len(managed_addons) == 0 {
+			if !d.HasChange("addons") {
+				var definedAddons []string
```
Comment (hasan4791) [inline ibm/service/kubernetes/resource_ibm_container_addons.go:540]: As discussed, we can reuse the existing addOnName variable here and get rid of Get & Set function and follow the existing flow.

File: ibm/service/kubernetes/resource_ibm_container_addons.go [hunk 5]
```diff
@@ -557,6 +574,20 @@ func resourceIBMContainerAddOnsUpdate(d *schema.ResourceData, meta interface{})
 			newList = new(schema.Set)
 		}
 		os := oldList.(*schema.Set)
+		// manage_all_addons was not previously set to false and the user has added additional or upgraded the existing addons, we set managed_addons to hold the old values
+		if !manageAllAddons && len(managed_addons) == 0 {
+			var oldAddons []string
```
Comment (hasan4791) [inline ibm/service/kubernetes/resource_ibm_container_addons.go:579]: Same comment as line no:540

**Review by hasan4791 — CHANGES_REQUESTED** [review]

File: ibm/service/kubernetes/resource_ibm_container_addons.go (hunk 3 above)
Comment (aayushsss1) [reply ibm/service/kubernetes/resource_ibm_container_addons.go:61]: An error comes up when we try to update tf, so I made it optional

Comment (hkantare) [reply ibm/service/kubernetes/resource_ibm_container_addons.go:61]: I think that not the fix for this if you mark it optioan user can provide it as input

File: ibm/service/kubernetes/resource_ibm_container_addons.go [hunk 6]
```diff
@@ -430,15 +430,13 @@ func resourceIBMContainerAddOnsRead(d *schema.ResourceData, meta interface{}) er
...
 	d.Set("cluster", cluster)
-	manageAllAddons := d.Get("manage_all_addons").(bool)
 	var addOns *schema.Set
-	if !manageAllAddons {
-
-		addOns, err = flattenAddOn(d, result)
-
-	} else {
-
+	manageAllAddons, ok := d.GetOkExists("manage_all_addons")
```
Comment (hasan4791) [inline ibm/service/kubernetes/resource_ibm_container_addons.go:434]: Use GetOk. GetOkExists is deprecated.

File: ibm/service/kubernetes/resource_ibm_container_addons.go [hunk 7]
```diff
@@ -430,15 +430,13 @@ func resourceIBMContainerAddOnsRead(d *schema.ResourceData, meta interface{}) er
...
+	if !ok || (ok && manageAllAddons.(bool)) {
 		addOns, err = flattenAddOns(result)
+		d.Set("managed_addons", nil)
```
Comment (hasan4791) [inline ibm/service/kubernetes/resource_ibm_container_addons.go:437]: Also set "manage_all_addons" parameter to true, if it doesn't exist.

**Review by hasan4791 — CHANGES_REQUESTED** [review]

**Review by hasan4791 — APPROVED** [review]

**Pull request merged**